crmprtd/insert.py

```diff
@@ -142,85 +178,133 @@ def bisect_insert_strategy(sesh, obs):
-        except IntegrityError as e:
-            log.debug(
-                "Failure, observation already exists",
-                extra={"obs": obs, "exception": e},
-            )
-            db_metrics = DBMetrics(0, 1, 0)
-        except InsertionError as e:
-            log.warning(
-                "Failure occured during insertion", extra={"obs": obs, "exception": e}
-            )
-            db_metrics = DBMetrics(0, 0, 1)
-        else:
-            log.debug("Success for single observation")
-            log.info("Successfully inserted observations: 1")
-            db_metrics = DBMetrics(1, 0, 0)
-        sesh.commit()
-        return db_metrics
-
+    elif len(observations) == 1:
+        return insert_single_obs(sesh, observations[0])
     # The happy case: add everything at once
     else:
         try:
             with sesh.begin_nested():
-                log.debug("New SAVEPOINT", extra={"num_obs": len(obs)})
-                sesh.add_all(obs)
+                log.debug("New SAVEPOINT", extra={"num_obs": len(observations)})
+                sesh.add_all(observations)
         except IntegrityError:
             log.debug("Failed, splitting observations.")
             sesh.rollback()
 
-            a, b = split(obs)
+            a, b = split(observations)
             dbm_a = bisect_insert_strategy(sesh, a)
             dbm_b = bisect_insert_strategy(sesh, b)
             return dbm_a + dbm_b
         else:
             log.info(
-                f"Successfully inserted observations: {len(obs)}",
-                extra={"num_obs": len(obs)},
+                f"Successfully inserted observations: {len(observations)}",
+                extra={"num_obs": len(observations)},
             )
-            db_metrics = DBMetrics(len(obs), 0, 0)
+            db_metrics = DBMetrics(len(observations), 0, 0)
         sesh.commit()
         return db_metrics
 
 
-def insert(sesh, observations, sample_size):
+def bulk_insert_strategy(sesh, observations):
+    """
+    This method performs a bulk insert of observations using the PostgreSQL dialect
+    INSERT ... ON CONFLICT DO IGNORE clause that allows bulk inserts with duplicates
+    to proceed without raising an exception (dups are ignored; new rows are inserted).
+
+    NOTE: This only works for PostgreSQL databases. Other dialects may support a
+    similar operation, but this code does not handle those cases.
+
+    :param sesh: SQLAlchemy database session
+    :param observations: List of observations to insert
+    :return: DMMetrics describing result of insertion
+    """
+    num_to_insert = len(observations)
+
+    # If you try to insert no rows with pg_insert, there's an unexpected insertion.
+    # Avoid that.
+    if num_to_insert == 0:
+        return DBMetrics(0, 0, 0)
+
+    try:
+        with sesh.begin_nested():
+            result = sesh.execute(
+                pg_insert(Obs)
+                .values(observations)
+                .on_conflict_do_nothing()
```

Comment:
Since we're not using the ORM mappings here for the insert, I have to wonder whether a significant part of the optimization is simply in the bypassing of the ORM and using raw SQL. E.g. when we create an instance of `Obs`, does that require a round trip (or multiple round trips) to the database to fill out the object attributes? If so, that would be the slow part, regardless of how we try to do the insert.

Reply:
Off the cuff, I don't think so. This is no more "raw SQL" than any other use of SQLAlchemy.

Not sure precisely what you mean by "create an instance of `Obs`". If you mean _within the code_, then it takes _zero_ round trips. If you mean _when an object is inserted in the database_, then it takes one round trip, as any other query would. Unless you emit a query, the database is not involved.